File: matiasmasso/Mat.NET/DT2/DTOImpagat.vb
Public Class DTOImpagat
    Inherits DTOBaseGuid

    Property csb As DTOCsb
    Property refBanc As String
    Property gastos As DTOAmt
    Property pagatACompte As DTOAmt
    Property fchAFP As Date
    Property fchSdo As Date
    'Property Insolvencia As Insolvencia
    Property status As StatusCodes
    Property asnefAlta As Date
    Property asnefBaixa As Date

    Property ccaIncobrable As DTOCca
    Property obs As String

    Property lastMemFch As Date

    Public Enum StatusCodes
        notSet
        enNegociacio
        conveni
        saldat
        insolvencia
    End Enum

    Public Enum OrderBy
        vto
        cliNomVto
    End Enum


    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Public ReadOnly Property Nominal As DTOAmt
        Get
            Dim retval As DTOAmt = Nothing
            If _Csb IsNot Nothing Then
                retval = _Csb.Amt
            End If
            Return retval
        End Get
    End Property

    Shared Function PendentAmbGastos(oImpagat As DTOImpagat) As DTOAmt
        Dim retval As DTOAmt = oImpagat.Nominal.Clone
        If oImpagat.PagatACompte IsNot Nothing Then
            retval = retval.Substract(oImpagat.PagatACompte)
        End If
        If oImpagat.Gastos IsNot Nothing Then
            Dim oGastos As DTOAmt = oImpagat.Gastos
            If oGastos.Eur < 40 Then oGastos = DTOAmt.Factory(40) 'minim despeses
            retval = retval.Add(oGastos)
        End If
        Return retval
    End Function

    Shared Function Pendent(oImpagat As DTOImpagat) As DTOAmt
        Dim retval As DTOAmt = oImpagat.Nominal
        If oImpagat.PagatACompte IsNot Nothing Then
            retval = retval.Substract(oImpagat.PagatACompte)
        End If
        Return retval
    End Function

    Shared Function GetGastos(oImpagat As DTOImpagat) As DTOAmt
        Dim oMinimGastos As DTOAmt = DTOAmt.Factory(40)
        Dim retval As DTOAmt = oImpagat.Gastos
        If retval Is Nothing Then
            retval = oMinimGastos
        Else
            If retval.Eur < oMinimGastos.Eur Then
                retval = oMinimGastos
            End If
        End If
        Return retval
    End Function
End Class


Public Class DTOImpagats
    Property Impagats As List(Of DTOImpagat)
    Property Cca As DTOCca
End Class

File: matiasmasso/Mat.NET/DTO/DTOImpagat.cs
using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOImpagat : DTOBaseGuid
    {
        public DTOCsb Csb { get; set; }
        public string RefBanc { get; set; }
        public DTOAmt Gastos { get; set; }
        public DTOAmt PagatACompte { get; set; }
        public DateTime FchAFP { get; set; }
        public DateTime FchSdo { get; set; }
        // Property Insolvencia As Insolvencia
        public StatusCodes Status { get; set; }
        public DateTime AsnefAlta { get; set; }
        public DateTime AsnefBaixa { get; set; }

        public DTOCca CcaIncobrable { get; set; }
        public string Obs { get; set; }

        public DateTime LastMemFch { get; set; }

        public enum StatusCodes
        {
            notSet,
            enNegociacio,
            conveni,
            saldat,
            insolvencia
        }

        public enum OrderBy
        {
            vto,
            cliNomVto
        }


        public DTOImpagat() : base()
        {
        }

        public DTOImpagat(Guid oGuid) : base(oGuid)
        {
        }

        public DTOAmt Nominal
        {
            get
            {
                DTOAmt retval = null/* TODO Change to default(_) if this is not a reference type */;
                if (Csb != null)
                    retval = Csb.Amt;
                return retval;
            }
        }

        public static DTOAmt PendentAmbGastos(DTOImpagat oImpagat)
        {
            DTOAmt retval = oImpagat.Nominal.Clone();
            if (oImpagat.PagatACompte != null)
                retval = retval.Substract(oImpagat.PagatACompte);
            if (oImpagat.Gastos != null)
            {
                DTOAmt oGastos = oImpagat.Gastos;
                if (oGastos.Eur < 40)
                    oGastos = DTOAmt.Factory(40); // minim despeses
                retval = retval.Add(oGastos);
            }
            return retval;
        }

        public static DTOAmt Pendent(DTOImpagat oImpagat)
        {
            DTOAmt retval = oImpagat.Nominal;
            if (oImpagat.PagatACompte != null)
                retval = retval.Substract(oImpagat.PagatACompte);
            return retval;
        }

        public static DTOAmt GetGastos(DTOImpagat oImpagat)
        {
            DTOAmt oMinimGastos = DTOAmt.Factory(40);
            DTOAmt retval = oImpagat.Gastos;
            if (retval == null)
                retval = oMinimGastos;
            else if (retval.Eur < oMinimGastos.Eur)
                retval = oMinimGastos;
            return retval;
        }
    }


    public class DTOImpagats
    {
        public List<DTOImpagat> Impagats { get; set; }
        public DTOCca Cca { get; set; }
    }
}
